Add 2.6 prebuilt models to Machine Learning (#2988)
opensearch-project/documentation-website: _ml-commons-plugin/pretrained-models.md

---
layout: default

> title: Pre-trained models

Reviewer: ```suggestion
title: Pretrained models
```

parent: Model-serving framework
nav_order: 120
---

# Pretrained models

> The model-serving framework supports a variety of open-source pretrained models that can assist with a variety of machine learning (ML) search and analytics use cases. 

Reviewer: ```suggestion
The model-serving framework supports a variety of open-source pretrained models that can assist with a range of machine learning (ML) search and analytics use cases. 
```

Reviewer: Avoiding the double use of "variety".

## Uploading pretrained models

To use a pretrained model in your OpenSearch cluster:

> 1. Select the model and download the model artifact and config.json file. For a list of pretrained models, see [supported pretrained model](#supported-pretrained-models).

Reviewer: In the link, should "model" be plural?

> 2. Upload the model using the [upload API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework#upload-model-to-opensearch). For more information on how to upload and use model, see [Model-serving framework]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework).

Reviewer: ```suggestion
2. Upload the model using the [upload API]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework#upload-model-to-opensearch). For more information on how to upload and use ML models, see [Model-serving framework]({{site.url}}{{site.baseurl}}/ml-commons-plugin/model-serving-framework).
```

## Supported pretrained models

> The model-serving framework supports the following models, categorized based on type. Although models with the same type will have similar use cases, each model has a different model size and performs differently depending on your cluster. For a comparison of the performances some pre-trained models, see the [sbert documentation](https://www.sbert.net/docs/pretrained_models.html#model-overview).

Reviewer: ```suggestion
The model-serving framework supports the following models, categorized by type. Although models with the same type will have similar use cases, each model has a different model size and performs differently depending on your cluster. For a comparison of the performances of some pretrained models, see the [sbert documentation](https://www.sbert.net/docs/pretrained_models.html#model-overview).
```

### Sentence transformers

Sentence transformer models map sentences and paragraphs across a dimensional dense vector space. The number of vectors depends on the model. Use these models for use cases such as clustering and semantic search.

> | **Model name** | **Vector dimensions** | **Torchscript artifact** | **ONNX artifact** |

Reviewer: We should introduce the table in the text that precedes it. "The following table provides..."

> | **Model name** | **Vector dimensions** | **Torchscript artifact** | **ONNX artifact** |

Reviewer 2: Is there a new standard that requires table headers to be bold? Should we be doing this across all docs?

Author: We don't have a standard one way or another. I always bold mine, so that the header labels are more obvious.

Since it's a formatting question, lets get a consensus and add it to the formatting guide.

Reviewer: Agreed. They should generally always be bold.

---
layout: default
title: Pretrained models
parent: Model-serving framework
nav_order: 120
---

# Pretrained models

> The model-serving framework supports a variety of open-source pretrained models that can assist with a range of machine learning (ML) search and analytics use cases. 

Reviewer 3: May be we can add:

We traced a few pre-trained models in our model repo for customers so that they can use these models in opensearch.

Author: Addressed this by adding a caveat to the sbert link.

## Uploading pretrained models

To use a pretrained model in your OpenSearch cluster:

> 1. Select the model and download the model artifact and config.json file. For a list of pretrained models, see [supported pretrained models](#supported-pretrained-models).

Reviewer 4: Upload models from our model repo is easy. User just need to provide model name, version and format. The supported model version starts from 1.0.1 

check https://github.com/opensearch-project/ml-commons/issues/745

```
POST /_plugins/_ml/models/_upload
{
  "name": "huggingface/sentence-transformers/all-MiniLM-L12-v2",
  "version": "1.0.1",
  "model_format": "TORCH_SCRIPT"
}
```

Author: How would the API know that we're uploading a pretrained model unless they download the artifact first? Does it come prepackaged with OpenSearch?

Reviewer 4: When upload model with "Bring your own model(BYOM)" way, user need to provide "url" in request parameter, for example
```
POST /_plugins/_ml/models/_upload
{
  "name": "all-MiniLM-L6-v2",
  "version": "1.0.0",
  "description": "test model",
  "model_format": "TORCH_SCRIPT",
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 384,
    "framework_type": "sentence_transformers",
    "all_config": "{\"_name_or_path\":\"nreimers/MiniLM-L6-H384-uncased\",\"architectures\":[\"BertModel\"],\"attention_probs_dropout_prob\":0.1,\"gradient_checkpointing\":false,\"hidden_act\":\"gelu\",\"hidden_dropout_prob\":0.1,\"hidden_size\":384,\"initializer_range\":0.02,\"intermediate_size\":1536,\"layer_norm_eps\":1e-12,\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6,\"pad_token_id\":0,\"position_embedding_type\":\"absolute\",\"transformers_version\":\"4.8.2\",\"type_vocab_size\":2,\"use_cache\":true,\"vocab_size\":30522}"
  },
  "url": "https://github.com/ylwu-amzn/ml-commons/blob/2.x_custom_m_helper/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/all-MiniLM-L6-v2_torchscript_sentence-transformer.zip?raw=true"
}
```
When upload pre-trained model from our model repo, user don't need to provide URL. If no URL provided, our code will try to construct pre-trained model URL with model name, version and format like this way
`https://artifacts.opensearch.org/models/ml-models/<model_name>/<version>/<model format>/<model_artifact_zip>`

If model name is `huggingface/sentence-transformers/msmarco-distilbert-base-tas-b`,  version is `1.0.1`, format is `onnx`, then model zip URL is  `https://artifacts.opensearch.org/models/ml-models/huggingface/sentence-transformers/msmarco-distilbert-base-tas-b/1.0.1/onnx/sentence-transformers_msmarco-distilbert-base-tas-b-1.0.0-torch_script.zip`

Reviewer 4: For pre-trained model, user done't need to download model zip and config
```suggestion
1. Select the model for a list of pretrained models, see the pre-trained model table below (add table link)
```

The model config json file is not readable, suggest link to the model table below.

> 1. Select the model and download the model artifact and config.json file. For a list of pretrained models, see [supported pretrained models](#supported-pretrained-models).

Reviewer 3: After `Select the model and download the model artifact and config.json file.` can we refer to the section `Sentence transformers`, so that customers don't get confused which models they need to download.

Author: @dhrubo-os: Since the plan is to add more models in the future, I would rather link to the general section.
